compiler/rustc_builtin_macros/src/assert/context.rs

```diff
@@ -0,0 +1,434 @@
+use rustc_ast::{
+    ptr::P,
+    token,
+    token::Nonterminal,
+    tokenstream::CanSynthesizeMissingTokens,
+    tokenstream::{DelimSpan, TokenStream, TokenTree},
+    Block, Expr, ExprKind, ItemKind, LitKind, MacArgs, MacCall, MacDelimiter, Path, PathSegment,
+    Stmt, StmtKind, StrStyle, StructRest, UseTree, UseTreeKind, DUMMY_NODE_ID,
+};
+use rustc_ast_pretty::pprust;
+use rustc_expand::base::ExtCtxt;
+use rustc_span::{
+    symbol::{sym, Ident, Symbol},
+    Span,
+};
+
+pub(super) struct Context<'cx, 'a> {
+    pub(super) cx: &'cx ExtCtxt<'a>,
+    pub(super) span: Span,
+    captures: Vec<Capture>,
+}
+
+impl<'cx, 'a> Context<'cx, 'a> {
+    pub(super) fn new(cx: &'cx ExtCtxt<'a>, span: Span) -> Self {
+        Self { cx, captures: Vec::new(), span }
+    }
+
+    /// Builds the whole `assert!` expression.
+    ///
+    /// {
+    ///    use ::core::asserting::*;
+    ///
+    ///    let mut __capture0 = Capture::Unevaluated;
+    ///    ...
+    ///
+    ///    if !{ ... } {
+    ///        panic!(
+    ///            "Assertion failed: ... \n With expansion: ...",
+    ///            (&__capture0).printer_kind().new(" ... ", __capture0),
+    ///            ...
+    ///        );
+    ///    }
+    /// }
+    pub(super) fn build(mut self, mut cond_expr: P<Expr>, panic_path: Path) -> P<Expr> {
+        let expr_str = escape_to_fmt(&pprust::expr_to_string(&cond_expr));
+        let fmt_str = &pprust::expr_to_string_cb(&cond_expr, |expr| {
+            if let ExprKind::Path(..) = expr.kind { Some("{:?}") } else { None }
+        });
+
+        self.manage_cond_expr(&mut cond_expr);
+        let glob_import = self.build_glob_import();
+        let panic = self.build_panic(&expr_str, &fmt_str, panic_path);
+
+        let Self { captures, cx, span, .. } = self;
+
+        let mut stmts = vec![glob_import];
+        stmts.extend(captures.into_iter().map(|c| c.decl));
+        stmts.push(cx.stmt_expr(cx.expr_if_not(span, cond_expr, panic, None)));
+        cx.expr_block(cx.block(span, stmts))
+    }
+
+    // Initial import to ease the construction of all `core::asserting` statements
+    //
+    // use ::core::asserting::*;
```

Comment:
It's better to make some helper function and use full paths everywhere instead of imports.
I'm not sure what shadowing issues these imports can cause in corner cases.